### ISSUE TITLE
RVM install fail on macOS Yosemite due expired curl's SSL certificate

### ISSUE THREAD

**Issue by bitaxis** [rvm/rvm#3886]
## Description

I have RVM 1.28.0 installed on Yosemite 10.10.5, but have been unable to install Rubies, as it ends with the following message:
> curl: (60) SSL certificate problem: certificate has expired

## Steps to reproduce

To reproduce, just run:

>$ rvm install 2.4.0

## Expected behavior

Obviously, I expected installation to succeed, like I've done many times before.

## Actual behavior

https://gist.github.com/bitaxis/a07327bc7366695d17d327e29f1fdd6f

An interesting thing I discovered by accident is that if I run ```rvm osx-ssl-certs update all``` immediately after encountering the failure to install, the the installation seems to resume and finishes successfully.

What gives?

By the way, running ```rvm install 2.4.0``` on a 2015 Retina MBP running El Capitan does not produce the same behavior, as it succeeds without issue.

As recommended by the guidelines, I also tried running ```rvm --debug install 2.4.0```.  The pertinent bits are here:

```
ruby-2.4.0 - #making binaries executable..
rubygems_setup version:latest rvm_ruby_string:ruby-2.4.0 rvm_ruby_home:/Users/nb/.rvm/rubies/ruby-2.4.0 rvm_ruby_binary:/Users/nb/.rvm/rubies/ruby-2.4.0/bin/ruby
Running(15): curl --fail --location --max-redirs 10 --max-time 1800 --connect-timeout 30 --retry-delay 2 --retry 3 -sS https://rubygems.org/api/v1/versions/rubygems-update.json
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html
```

## Environment info

```
ruby-2.3.1:

  system:
    uname:       "Darwin lightning 14.5.0 Darwin Kernel Version 14.5.0: Sun Sep 25 22:07:15 PDT 2016; root:xnu-2782.50.9~1/RELEASE_X86_64 x86_64"
    system:      "osx/10.10/x86_64"
    bash:        "/bin/bash => GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin14)"
    zsh:         "/bin/zsh => zsh 5.0.5 (x86_64-apple-darwin14.0)"

  rvm:
    version:      "rvm 1.28.0 (latest) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]"
    updated:      "41 minutes 9 seconds ago"
    path:         "/Users/nb/.rvm"

  ruby:
    interpreter:  "ruby"
    version:      "2.3.1p112"
    date:         "2016-04-26"
    platform:     "x86_64-darwin14"
    patchlevel:   "2016-04-26 revision 54768"
    full_version: "ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin14]"

  homes:
    gem:          "/Users/nb/.rvm/gems/ruby-2.3.1"
    ruby:         "/Users/nb/.rvm/rubies/ruby-2.3.1"

  binaries:
    ruby:         "/Users/nb/.rvm/rubies/ruby-2.3.1/bin/ruby"
    irb:          "/Users/nb/.rvm/rubies/ruby-2.3.1/bin/irb"
    gem:          "/Users/nb/.rvm/rubies/ruby-2.3.1/bin/gem"
    rake:         "/Users/nb/.rvm/rubies/ruby-2.3.1/bin/rake"

  environment:
    PATH:         "/Users/nb/.rvm/gems/ruby-2.3.1/bin:/Users/nb/.rvm/gems/ruby-2.3.1@global/bin:/Users/nb/.rvm/rubies/ruby-2.3.1/bin:/Users/nb/.rvm/bin:.:/Users/nb/bin:/opt/local/bin:/opt/local/sbin:/Users/nb/.nvm/versions/node/v7.3.0/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
    GEM_HOME:     "/Users/nb/.rvm/gems/ruby-2.3.1"
    GEM_PATH:     "/Users/nb/.rvm/gems/ruby-2.3.1:/Users/nb/.rvm/gems/ruby-2.3.1@global"
    MY_RUBY_HOME: "/Users/nb/.rvm/rubies/ruby-2.3.1"
    IRBRC:        "/Users/nb/.rvm/rubies/ruby-2.3.1/.irbrc"
    RUBYOPT:      ""
    gemset:       ""
```

**Comment by bitaxis** [comment]
After much digging, the following hack enables me to install Rubies with RVM once again:

```bash
$ cd /opt/local/share/curl
$ sudo wget https://curl.haxx.se/ca/cacert.pem
$ sudo mv curl-ca-bundle.crt curl-ca-bundle.crt-original
$ sudo mv cacert.pem curl-ca-bundle.crt
$ rvm remove 2.4.0
$ rvm install 2.4.0 # some time later...
$ rvm use 2.4.0
$ ruby --version
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-darwin14]
$ rvm gemset list
gemsets for ruby-2.4.0 (found in /Users/nb/.rvm/gems/ruby-2.4.0)
=> (default)
   global
```

**Comment by pkuczynski** [comment]
@bitaxis I prepared patch for you. Could you revert your changes and try again with following branch:

```
rvm get branch /bugfix/3886-curl-invalid-ssl-cert
rvm remove 2.4.0
rvm install 2.4.0
rvm use 2.4.0
ruby --version
rvm gemset list
```

**Comment by pkuczynski** [comment]
@bitaxis would you find some time to test it?

**Comment by bitaxis** [comment]
Sure.  I'll try to do so in the next few days.

**Comment by bitaxis** [comment]
@pkuczynski, I rolled back my hack, verified that ```rvm install 2.4.0``` still resulted in expired SSL certificate, and tried your patch.

Unforunately, the result is the same.  I still get the same error...

**Comment by pkuczynski** [comment]
Thanks for helping me with this.

I forgot to mention that you need to run `rvm osx-ssl-certs update`. Sorry for that. After that check if `/opt/local/share/curl/curl-ca-bundle.crt` has been updated.

If it has not, please get the latest patch again, as I added some more debugging information and post the output.

```
rvm get branch /bugfix/3886-curl-invalid-ssl-cert
rvm osx-ssl-certs update --debug
rvm remove 2.4.0
rvm install 2.4.0
```

Thanks for all your help on making RVM better!

**Comment by bitaxis** [comment]
I rolled back my hack again and ran these commands:

```bash
$ rvm get branch /bugfix/3886-curl-invalid-ssl-cert
$ rvm osx-ssl-certs update --debug
```

The second command yielded the following output:

```
$ rvm osx-ssl-certs update --debug
Expected cURL certificate bundle location: /opt/local/share/curl/curl-ca-bundle.crt
ruby-2.3.1 # getting ruby certs path
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
/opt/local/etc/openssl/cert.pem # updating certs
Updating certificates bundle /opt/local/etc/openssl/cert.pem: Already up to date.
Updating cURL certificate bundle /opt/local/share/curl/curl-ca-bundle.crt
Updating certificates bundle /opt/local/share/curl/curl-ca-bundle.crt: Already up to date.
```

I checked ```/opt/local/share/curl/curl-ca-bundle.crt```, and it was unmodified.  I even ```diff```'d the file against a copy from before I ran the commands, and no differences were found.

**Comment by pkuczynski** [comment]
Ok, I made some further changes. Could you please run:

```
rvm get branch /bugfix/3886-curl-invalid-ssl-cert
rvm osx-ssl-certs all update --debug
rvm remove 2.4.0
rvm install 2.4.0
```

If if does not work you can also try:

```
rvm osx-ssl-certs all update --debug --force
```

Please post both gists if any of the above failed.

**Comment by pkuczynski** [comment]
@bitaxis any chance to get this tested (hopefully the last time)?

**Comment by bitaxis** [comment]
@pkuczynski,

I ran

```bash
$ rvm get branch /bugfix/3886-curl-invalid-ssl-cert
# ...
$ rvm osx-ssl-certs update all --debug # proper syntax
Selected SSL certs for: ruby-2.2.3
Getting ruby certs path for ruby-2.2.3
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting ruby certs path for ruby-2.3.0
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting ruby certs path for ruby-2.3.1
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting ruby certs path for ruby-2.4.0
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting SSL certs path for curl
Expected cURL certificate bundle location: /opt/local/share/curl/curl-ca-bundle.crt
Found cURL certificate bundle /opt/local/share/curl/curl-ca-bundle.crt
Updating SSL certs /etc/openssl/cert.pem
Updating certificates bundle /etc/openssl/cert.pem: Already up to date.
Updating SSL certs /opt/local/etc/openssl/cert.pem
Updating certificates bundle /opt/local/etc/openssl/cert.pem: Already up to date.
Updating SSL certs /opt/local/share/curl/curl-ca-bundle.crt
Updating certificates bundle /opt/local/share/curl/curl-ca-bundle.crt: Already up to date.
__rvm_rm_rf already gone: /Users/nb/.rvm/tmp/2786*
```

But the file ```/opt/local/share/curl/curl-ca-bundle.crt``` didn't get touched at all.

Then I ran the variant with the ```--force``` option:

```bash
$ rvm osx-ssl-certs update all --debug --force
Selected SSL certs for: ruby-2.2.3
Getting ruby certs path for ruby-2.2.3
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting ruby certs path for ruby-2.3.0
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting ruby certs path for ruby-2.3.1
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting ruby certs path for ruby-2.4.0
Running /Users/nb/.rvm/hooks/after_use_jruby
Running /Users/nb/.rvm/hooks/after_use_maglev
Getting SSL certs path for curl
Expected cURL certificate bundle location: /opt/local/share/curl/curl-ca-bundle.crt
Found cURL certificate bundle /opt/local/share/curl/curl-ca-bundle.crt
Updating SSL certs /etc/openssl/cert.pem
Updating certificates bundle /etc/openssl/cert.pem: Updating certificates bundle '/etc/openssl/cert.pem'
Updated.
Updating SSL certs /opt/local/etc/openssl/cert.pem
Updating certificates bundle /opt/local/etc/openssl/cert.pem: Updating certificates bundle '/opt/local/etc/openssl/cert.pem'
Updated.
Updating SSL certs /opt/local/share/curl/curl-ca-bundle.crt
Updating certificates bundle /opt/local/share/curl/curl-ca-bundle.crt: Updating certificates bundle '/opt/local/share/curl/curl-ca-bundle.crt'
Updated.
__rvm_rm_rf already gone: /Users/nb/.rvm/tmp/2786*
```
This time, the file ```/opt/local/share/curl/curl-ca-bundle.crt``` got touched, but the content is identical.

Then I ran rest of the commands, but to no avail:

```bash
$ rvm remove 2.4.0
ruby-2.4.0 - #removing src/ruby-2.4.0..
ruby-2.4.0 - #removing rubies/ruby-2.4.0..
ruby-2.4.0 - #removing gems....
ruby-2.4.0 - #removing wrappers....
ruby-2.4.0 - #removing environments....
Using /Users/nb/.rvm/gems/ruby-2.3.1
nb@lightning /opt/local/share/curl
(13:32) $ rvm install 2.4.0
Searching for binary rubies, this might take some time.
No binary rubies available for: osx/10.10/x86_64/ruby-2.4.0.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for osx_port.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Warning: port definitions are more than two weeks old, consider updating them by running 'port selfupdate'.
Certificates bundle '/opt/local/etc/openssl/cert.pem' is already up to date.
Requirements installation successful.
Installing Ruby from source to: /Users/nb/.rvm/rubies/ruby-2.4.0, this may take a while depending on your cpu(s)...
ruby-2.4.0 - #downloading ruby-2.4.0, this may take a while depending on your connection...
ruby-2.4.0 - #extracting ruby-2.4.0 to /Users/nb/.rvm/src/ruby-2.4.0....
ruby-2.4.0 - #configuring...................................................................
ruby-2.4.0 - #post-configuration.
ruby-2.4.0 - #compiling.............................................................
ruby-2.4.0 - #installing.......
ruby-2.4.0 - #making binaries executable..
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
There was an error while trying to resolve rubygems version for 'latest'. 
Halting the installation.
```
